File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/security/warrants/SecurityWarrantDelegationStateMachine.java
package com.foundgine.core.semantic.security.warrants;
import java.util.Map; import java.util.concurrent.ConcurrentHashMap;
public final class SecurityWarrantDelegationStateMachine {
 public enum State { ACTIVE, REVOKED, COMPROMISED }
 public record Snapshot(String warrantId,String warrantDigest,State state,long sequence,String activeKeyId) {}
 public record Transition(Snapshot before,Snapshot after,String operation) {}
 private static final class Cell { final Object gate=new Object(); State state=State.ACTIVE; long sequence; String activeKeyId; }
 private final Map<String,Cell> cells=new ConcurrentHashMap<>();
 public Snapshot register(SecurityWarrant w,String activeKeyId){if(w==null)throw new NullPointerException();var c=cells.computeIfAbsent(key(w),k->new Cell());synchronized(c.gate){if(c.sequence!=0)throw new IllegalStateException("Warrant is already registered in the delegation state machine.");c.sequence=1;c.activeKeyId=activeKeyId;return snapshot(w,c);}}
 public Transition revoke(SecurityWarrant w){return transition(w,"revoke",s->s==State.ACTIVE,c->c.state=State.REVOKED);}
 public Transition compromise(SecurityWarrant w){return transition(w,"compromise",s->s!=State.COMPROMISED,c->c.state=State.COMPROMISED);}
 public Transition rotateKey(SecurityWarrant w,String newKeyId){if(newKeyId==null||newKeyId.isBlank())throw new IllegalArgumentException("Key id is required.");return transition(w,"rotate-key",s->s==State.ACTIVE,c->c.activeKeyId=newKeyId);}
 public Snapshot assertCanDelegate(SecurityWarrant w){var c=get(w);synchronized(c.gate){if(c.state!=State.ACTIVE)throw new IllegalStateException("A revoked or compromised warrant cannot delegate.");return snapshot(w,c);}}
 public Snapshot read(SecurityWarrant w){var c=get(w);synchronized(c.gate){return snapshot(w,c);}}
 private Transition transition(SecurityWarrant w,String op,java.util.function.Predicate<State> allowed,java.util.function.Consumer<Cell> apply){if(w==null)throw new NullPointerException();var c=get(w);synchronized(c.gate){var b=snapshot(w,c);if(!allowed.test(c.state))throw new IllegalStateException("Illegal delegation state transition '"+op+"' from "+c.state+".");apply.accept(c);c.sequence++;return new Transition(b,snapshot(w,c),op);}}
 private Cell get(SecurityWarrant w){var c=cells.get(key(w));if(c==null)throw new IllegalStateException("Warrant is not registered in the delegation state machine.");return c;}private static Snapshot snapshot(SecurityWarrant w,Cell c){return new Snapshot(w.id(),w.digest(),c.state,c.sequence,c.activeKeyId);}private static String key(SecurityWarrant w){return w.id()+"\u001f"+w.digest();}
}

File: src/csharp/Foundgine.Core/Semantic/Security/Warrants/SecurityWarrantDelegationStateMachine.cs
namespace Foundgine.Core.Semantic.Security.Warrants;

/// <summary>Linearizable lifecycle states for delegation authority.</summary>
public enum SecurityWarrantDelegationState
{
    Active = 0,
    Revoked = 1,
    Compromised = 2
}

/// <summary>Immutable state observed at one linearization point.</summary>
public sealed record SecurityWarrantDelegationStateSnapshot(
    string WarrantId,
    string WarrantDigest,
    SecurityWarrantDelegationState State,
    long Sequence,
    string? ActiveKeyId);

/// <summary>Atomic state transition result.</summary>
public sealed record SecurityWarrantDelegationTransition(
    SecurityWarrantDelegationStateSnapshot Before,
    SecurityWarrantDelegationStateSnapshot After,
    string Operation);

/// <summary>
/// Small, provider-neutral state machine used to define the legal lifecycle boundary.
/// The lock is the model of the database serialization boundary; production stores must
/// provide an equivalent atomic compare-and-transition operation.
/// </summary>
public sealed class SecurityWarrantDelegationStateMachine
{
    private sealed class Cell
    {
        public readonly object Gate = new();
        public SecurityWarrantDelegationState State = SecurityWarrantDelegationState.Active;
        public long Sequence;
        public string? ActiveKeyId;
    }

    private readonly System.Collections.Concurrent.ConcurrentDictionary<string, Cell> _cells =
        new(StringComparer.Ordinal);

    public SecurityWarrantDelegationStateSnapshot Register(SecurityWarrant warrant, string? activeKeyId = null)
    {
        ArgumentNullException.ThrowIfNull(warrant);
        var cell = _cells.GetOrAdd(Key(warrant), _ => new Cell { ActiveKeyId = activeKeyId });
        lock (cell.Gate)
        {
            if (cell.Sequence != 0)
                throw new InvalidOperationException("Warrant is already registered in the delegation state machine.");
            cell.Sequence = 1;
            cell.ActiveKeyId = activeKeyId;
            return Snapshot(warrant, cell);
        }
    }

    public SecurityWarrantDelegationTransition Revoke(SecurityWarrant warrant)
        => Transition(warrant, "revoke", static s => s == SecurityWarrantDelegationState.Active,
            static cell => cell.State = SecurityWarrantDelegationState.Revoked);

    public SecurityWarrantDelegationTransition Compromise(SecurityWarrant warrant)
        => Transition(warrant, "compromise", static s => s != SecurityWarrantDelegationState.Compromised,
            static cell => cell.State = SecurityWarrantDelegationState.Compromised);

    /// <summary>Rotates the issuer key without changing authority state.</summary>
    public SecurityWarrantDelegationTransition RotateKey(SecurityWarrant warrant, string newKeyId)
    {
        if (string.IsNullOrWhiteSpace(newKeyId)) throw new ArgumentException("Key id is required.", nameof(newKeyId));
        return Transition(warrant, "rotate-key", static s => s == SecurityWarrantDelegationState.Active,
            cell => cell.ActiveKeyId = newKeyId);
    }

    /// <summary>
    /// Re-delegation is allowed only from an active, non-compromised parent and therefore
    /// must observe the same linearization point as revocation/compromise/key rotation.
    /// </summary>
    public SecurityWarrantDelegationStateSnapshot AssertCanDelegate(SecurityWarrant parent)
    {
        ArgumentNullException.ThrowIfNull(parent);
        var cell = Get(parent);
        lock (cell.Gate)
        {
            if (cell.State != SecurityWarrantDelegationState.Active)
                throw new InvalidOperationException("A revoked or compromised warrant cannot delegate.");
            return Snapshot(parent, cell);
        }
    }

    public SecurityWarrantDelegationStateSnapshot Read(SecurityWarrant warrant)
    {
        ArgumentNullException.ThrowIfNull(warrant);
        var cell = Get(warrant);
        lock (cell.Gate) return Snapshot(warrant, cell);
    }

    private SecurityWarrantDelegationTransition Transition(
        SecurityWarrant warrant,
        string operation,
        Func<SecurityWarrantDelegationState, bool> allowed,
        Action<Cell> apply)
    {
        ArgumentNullException.ThrowIfNull(warrant);
        var cell = Get(warrant);
        lock (cell.Gate)
        {
            var before = Snapshot(warrant, cell);
            if (!allowed(cell.State))
                throw new InvalidOperationException(
                    $"Illegal delegation state transition '{operation}' from {cell.State}.");
            apply(cell);
            cell.Sequence = checked(cell.Sequence + 1);
            var after = Snapshot(warrant, cell);
            return new(before, after, operation);
        }
    }

    private Cell Get(SecurityWarrant warrant)
    {
        var key = Key(warrant);
        if (!_cells.TryGetValue(key, out var cell))
            throw new InvalidOperationException("Warrant is not registered in the delegation state machine.");
        return cell;
    }

    private static SecurityWarrantDelegationStateSnapshot Snapshot(SecurityWarrant warrant, Cell cell) =>
        new(warrant.Id, warrant.Digest, cell.State, cell.Sequence, cell.ActiveKeyId);

    private static string Key(SecurityWarrant warrant) => warrant.Id + "\u001f" + warrant.Digest;
}